parlai/tasks/light_genderation_bias/agents.py

```diff
@@ -0,0 +1,405 @@
+            help='Append the conditional bucket to the end of the input or not',
+        )
+        parser.add_argument(
+            '--add-counterfactual',
+            type='bool',
+            default=True,
+            help='Load counterfactual data for training',
+        )
+        parser.add_argument(
+            '--force-conditional',
+            type=str,
+            default='none',
+            help='Add one bucket to everything, regardless of the true bin',
+        )
+        parser.add_argument(
+            '--bucket-only',
+            type=str,
+            default='none',
+            help='Only train/evaluate on examples from a specific bucket',
+        )
+
+        parser.set_defaults(
+            light_use_objects=False,
+            light_use_setting=False,
+            light_use_affordances=False,
+            light_use_persona_names=False,
+            light_use_persona='none',
+            light_use_action='none',
+            light_use_emote='none',
+            use_reply='none',
+        )
+        return parser
+
+    def __init__(self, opt, shared=None):
+        self.opt = opt
+        self.gender_dict = read_gender_tsv(os.path.join(_path(opt), GENDERED_LIST))
+        self.fixed_random = random.Random(42)
+
+        self.add_conditional = opt['add_conditional']
+        self.add_counterfactual = opt['add_counterfactual']
+        self.add_new_data = opt['add_new_data']
+
+        self.force_conditional = opt['force_conditional']
+        self.bucket_only = opt['bucket_only']
+        if self.force_conditional == 'none':
+            self.force_conditional = None
+        if self.bucket_only == 'none':
+            self.bucket_only = None
+
+        if shared and 'data' in shared:
+            self.data = shared['data']
+        else:
+            self.data = self._setup_data(opt)
+
+        super().__init__(opt, shared)
+        self.reset()
+
+    def num_episodes(self):
+        return len(self.data)
+
+    def num_examples(self):
+        return len(self.data)
+
+    def get_bucket(self, text):
+        """
+        Get the bin or "bucket" that a string falls into.
+
+        The bucket is determined based on the presence of male or femael gendered words
+        in the text.
+        """
+        text = format_text(text)
+        # we're bucketing based on the number of gendered words
+        f_count, m_count, n_count = get_finegrained_count(text, self.gender_dict)
+        # get f bucket
+        if f_count == 0:
+            f_bucket = 'f0'
+        else:
+            f_bucket = 'f1'
+
+        # get m bucket
+        if m_count == 0:
+            m_bucket = 'm0'
+        else:
+            m_bucket = 'm1'
+
+        # combine bucket
+        all_bucket = ' '.join([f_bucket, m_bucket])
+        return all_bucket
+
+    def _flip_ex(self, ex):
+        """
+        Return the counterfactual example for a given example (i.e. swap 'he' --> 'she')
+        """
+
+        def flip_str(txt_lines):
```

Comment:
lot of indentation here huh. Maybe consider pulling this out instead of a sub function? doesn't matter, it's just a little daunting.